internal/statemachine/classic_states.go

```diff
@@ -70,20 +70,43 @@ func (stateMachine *StateMachine) parseImageDefinition() error {
 	}
 
 	// do custom validation for gadgetURL being required if gadget is not pre-built
-	if imageDefinition.Gadget.GadgetType != "prebuilt" && imageDefinition.Gadget.GadgetURL == "" {
-		jsonContext := gojsonschema.NewJsonContext("gadget_validation", nil)
-		errDetail := gojsonschema.ErrorDetails{
-			"key":   "gadget:type",
-			"value": imageDefinition.Gadget.GadgetType,
-		}
-		result.AddError(
-			imagedefinition.NewMissingURLError(
-				gojsonschema.NewJsonContext("missingURL", jsonContext),
-				52,
+	if imageDefinition.Gadget != nil {
+		if imageDefinition.Gadget.GadgetType != "prebuilt" && imageDefinition.Gadget.GadgetURL == "" {
+			jsonContext := gojsonschema.NewJsonContext("gadget_validation", nil)
+			errDetail := gojsonschema.ErrorDetails{
+				"key":   "gadget:type",
+				"value": imageDefinition.Gadget.GadgetType,
+			}
+			result.AddError(
+				imagedefinition.NewMissingURLError(
+					gojsonschema.NewJsonContext("missingURL", jsonContext),
+					52,
+					errDetail,
+				),
 				errDetail,
-			),
-			errDetail,
-		)
+			)
+		}
+	}
+
+	// don't allow any images to be created without a gadget
+	if imageDefinition.Gadget == nil {
+		if imageDefinition.Artifacts.Img == nil ||
```

Comment:
I'm thinking, for the future, if there's a way of making this check better. Since I'm afraid we'd just forget, when we add new artifact types, to add the new artifact to this list here. I don't have any good ideas though, since we'd have to somehow nicely group all the `Artifacts` that need the gadget. This is non-blocking though.

Reply:
I've pushed an update to make this better.